### PR TITLE
Optimize speed

### PR DESCRIPTION
**Pull request by sztanko**
Because of 3c8550f106f583506151914485cdbc21969e0070 the performance of cubedb get method has decreased - it has not do unnecessary autoboxing when incrementing on top of extra hashmap lookup instead of array.
This PR introduces separate strategies for different cases, also uses Trove Object -> long hashmap instead of HashMap<Object, Long>.

### CODE REVIEW

File: src/main/java/org/cubedb/core/counters/ArrayBasedMultiDimensionalCounter.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+				this.values[i][j] = new long[thirdDimSizes[j]];
+			}
+	}
+
+	@Override
+	public void add(long value, final int... fields) {
+		final int f0 = fields[0];
+		final int f1 = fields[1];
+		final int f2 = fields[2];
+		values[f0][f1][f2]+=value;
```
Comment (vkazanov) [inline src/main/java/org/cubedb/core/counters/ArrayBasedMultiDimensionalCounter.java]: whitespaces around the += operator are missing

File: src/main/java/org/cubedb/core/counters/ArrayBasedMultiDimensionalCounter.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	public ArrayBasedMultiDimensionalCounter(int firstDimSize, int secondDimSize, int... thirdDimSizes) {
+		this.values = new long[firstDimSize][secondDimSize][];
+		for (int i = 0; i < firstDimSize; i++)
+			for (int j = 0; j < secondDimSize; j++) {
+				this.values[i][j] = new long[thirdDimSizes[j]];
+			}
+	}
+
+	@Override
+	public void add(long value, final int... fields) {
```
Comment (vkazanov) [inline src/main/java/org/cubedb/core/counters/ArrayBasedMultiDimensionalCounter.java]: why do you need an array here? You know the number of dimensions here.

File: src/main/java/org/cubedb/core/counters/CounterContainer.java [hunk 3]
```diff
@@ -1,96 +1,75 @@
...
 import java.util.Map;
-import java.util.HashMap;
-import java.util.Arrays;
 
 import org.cubedb.core.lookups.Lookup;
 import org.cubedb.core.beans.GroupedSearchResultRow;
 import org.cubedb.core.beans.SearchResult;
 
-public class CounterContainer {
+public class CounterContainer{
```
Comment (vkazanov) [inline src/main/java/org/cubedb/core/counters/CounterContainer.java]: whitespace missing after the class name

File: src/main/java/org/cubedb/core/counters/CounterContainer.java [hunk 4]
```diff
@@ -1,96 +1,75 @@
...
-		this.lookups = lookups;
-		sideCounters = new HashMap<MultiKey, Long>();
+		this.lookups = lookups;		
 	}
 
 	public void initSideCounters() {
 		isGroupLookup = false;
 		groupFieldName = SearchResult.FAKE_GROUP_FIELD_NAME;
+		int[] fieldSizes= new int[fieldLookup.size()];
+		for(int i=0;i<fieldSizes.length;i++)
```
Comment (vkazanov) [inline src/main/java/org/cubedb/core/counters/CounterContainer.java]: whitespaces

**Comment by vkazanov** [comment]
...notice how there are not real changes in OffheapPartition

File: src/main/java/org/cubedb/core/counters/QuartIntLongConsumer.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package org.cubedb.core.counters;
+
+public interface QuartIntLongConsumer {
```
Comment (vkazanov) [inline src/main/java/org/cubedb/core/counters/QuartIntLongConsumer.java]: I don't see any code using this interface.

Comment (sztanko) [inline src/main/java/org/cubedb/core/counters/QuartIntLongConsumer.java]: It's implemented as lambdas

**Pull request merged**